src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,588 @@
+		Preconditions.checkNotNull(function, "function is null");
+		Skript.debug("Registering function '" + function.getName() + "'");
+
+		String name = function.getName();
+		if (!name.matches(FUNCTION_NAME_PATTERN)) {
+			throw new SkriptAPIException("Invalid function name '" + name + "'");
+		}
+
+		// namespace
+		Namespace ns = GLOBAL_NAMESPACE;
+		if (namespace != null && function.getSignature().isLocal()) {
+			ns = new Namespace(Scope.LOCAL, namespace);
+		}
+
+		FunctionIdentifier identifier = FunctionIdentifier.of(function.getSignature());
+		if (!signatureExists(ns, identifier)) {
+			register(namespace, function.getSignature());
+		}
+
+		// register
+		Map<FunctionIdentifier, Function<?>> identifierToFunction = functions.getOrDefault(ns, new HashMap<>());
+		Function<?> existing = identifierToFunction.put(identifier, function);
+		if (existing != null) {
+			alreadyRegisteredError(name, identifier, ns);
+		}
+
+		functions.put(ns, identifierToFunction);
+	}
+
+	private static void alreadyRegisteredError(String name, FunctionIdentifier identifier, Namespace namespace) {
+		throw new SkriptAPIException("Function '%s' with parameters %s is already registered in %s"
+			.formatted(name, Arrays.toString(Arrays.stream(identifier.args).map(Class::getSimpleName).toArray()),
+				namespace));
+	}
+
+	/**
+	 * Checks if a signature with the given name and arguments exists in the given namespace.
+	 *
+	 * <ul>
+	 * <li>If {@code namespace} is null, only global signatures will be checked.</li>
+	 * <li>If {@code args} is null or empty,
+	 * the first function with the same name as the {@code name} param will be returned.</li>
+	 * </ul>
+	 *
+	 * @param namespace The namespace to check in.
+	 *                  Usually represents the path of the script a signature is registered in.
+	 * @param name      The name of the function.
+	 * @param args      The types of the arguments of the function.
+	 * @return True if a signature with the given name and argument types exists in the script, false otherwise.
+	 */
+	public static boolean signatureExists(@Nullable String namespace, @NotNull String name, Class<?>... args) {
+		if (namespace == null) {
+			return signatureExists(GLOBAL_NAMESPACE, FunctionIdentifier.of(name, false, args));
+		}
+
+		return signatureExists(new Namespace(Scope.LOCAL, namespace.toLowerCase()), FunctionIdentifier.of(name, true, args));
+	}
+
+	/**
+	 * Checks if a function with the given name and arguments exists in the namespace.
+	 *
+	 * @param namespace  The namespace to check in.
+	 * @param identifier The identifier of the function.
+	 * @return True if a function with the given name and arguments exists in the namespace, false otherwise.
+	 */
+	private static boolean signatureExists(@NotNull Namespace namespace, @NotNull FunctionIdentifier identifier) {
+		Preconditions.checkNotNull(namespace, "namespace is null");
+		Preconditions.checkNotNull(identifier, "identifier is null");
```

Comment:
all of these preconditions should explain what the conditions are, not what the current state is. See earlier comment about `cannot be` vs `is`